src/ai/backend/manager/idle.py

```diff
@@ -307,24 +318,196 @@ class UtilizationIdleChecker(BaseIdleChecker):
+        )
+
+        log.info(
+            f"UtilizationIdleChecker(%): "
+            f"cpu({self.resource_thresholds['cpu_threshold']}), "
+            f"mem({self.resource_thresholds['mem_threshold']}), "
+            f"cuda({self.resource_thresholds['cuda_threshold']}), "
+            f"cuda.mem({self.resource_thresholds['cuda_mem_threshold']}), "
+            f"resource availablity string: {self.resource_avail}, "
+            f"threshold check condition: {self.threshold_condition}, "
+            f"moving window: {self.window.total_seconds()}s"
+        )
 
     async def update_app_streaming_status(
         self,
         session_id: SessionId,
         status: AppStreamingStatus,
     ) -> None:
-        pass
+        if status == AppStreamingStatus.HAS_ACTIVE_CONNECTIONS:
+            await self._disable_timeout(session_id)
+        elif status == AppStreamingStatus.NO_ACTIVE_CONNECTIONS:
+            await self._update_timeout(session_id)
+
+    async def _disable_timeout(self, session_id: SessionId) -> None:
+        log.debug(f"UtilizationIdleChecker._disable_timeout({session_id})")
+        await self._redis.set(
+            f"session.{session_id}.last_execution", "0", exist=self._redis.SET_IF_EXIST
+        )
+
+    async def _update_timeout(self, session_id: SessionId) -> None:
+        log.debug(f"UtilizationIdleChecker._update_timeout({session_id})")
+        t = await self._redis.time()
+        await self._redis.set(
+            f"session.{session_id}.last_execution",
+            f"{t:.06f}",
+            expire=max(86400, int(self.window.total_seconds() * 2)),
+        ),
+
+    async def _session_started_cb(
+        self,
+        context: None,
+        source: AgentId,
+        event: SessionStartedEvent,
+    ) -> None:
+        await self._update_timeout(event.session_id)
+
+    async def _execution_started_cb(
+        self,
+        context: None,
+        source: AgentId,
+        event: ExecutionStartedEvent,
+    ) -> None:
+        await self._disable_timeout(event.session_id)
+
+    async def _execution_exited_cb(
+        self,
+        context: None,
+        source: AgentId,
+        event: ExecutionFinishedEvent | ExecutionTimeoutEvent | ExecutionCancelledEvent,
+    ) -> None:
+        await self._update_timeout(event.session_id)
+
+    async def _do_idle_check(
+        self,
+        context: None,
+        source: AgentId,
+        event: DoIdleCheckEvent,
+    ) -> None:
+
+        try:
+            return await super()._do_idle_check(context, source, event)
+        finally:
+            pass
+
+    async def check_session(self, session: Row, dbconn: SAConnection) -> bool:
+        session_id = session["id"]
+        active_streams = await self._redis.zcount(
+            f"session.{session_id}.active_app_connections"
+        )
+        if active_streams is not None and active_streams > 0:
+            return True
+
+        redis = await aioredis.create_redis("redis://127.0.0.1:8111/")
```

Comment:
I am having a hard time understanding this code. Obviously, the hard-coded address should be changed to a config item, but I am curious why is it needed to create a Redis connection when `BaseIdleChecker` has already created one in `__ainit__`?

Reply:
Redis connection based on address in config item doesn't return any values. The url for the config redis address is _redis://127.0.0.1:8111/2_
I needed to work with actual data to move forward, so made redis connection with the appropriate url address.

Reply:
In that case, you should use [DB number 0 (STAT_DB)](https://github.com/lablup/backend.ai-manager/blob/main/src/ai/backend/manager/defs.py#L27), so the hard-coded string should be replaced something like:

```python
from ai.backend.manager.defs import REDIS_STAT_DB
...
await aioredis.create_redis(
    str(self._shared_config.get_redis_url(db=REDIS_STAT_DB))
)
```